### PR TITLE
Add validation: warn when pad_token_id is in eos_token_id list

### LINKED ISSUE #48016: GenerationConfig should validate that pad_token_id is not in eos_token_id list
## Problem

When a model is fine-tuned with TRL's SFTTrainer, the saved `generation_config.json` can end up with `pad_token_id` set to a token that is also in the `eos_token_id` list. This causes `model.generate()` to emit EOS immediately, producing empty completions.

## Reproduction

We fine-tuned `nvidia/NVIDIA-Nemotron-Nano-9B-v2` with TRL SFTTrainer. The base model has:
```json
"pad_token_id": 0,
"eos_token_id": 12
```

After SFT training, the saved `generation_config.json` had:
```json
{
  "eos_token_id": [2, 11, 12],
  "pad_token_id": 12
}
```

Token 12 (`<|im_end|>`) is both pad AND eos. When `model.generate()` is called with `pad_token_id=12`, the model sees token 12 as EOS and stops immediately. HumanEval scored **0.0% (0/164)** — every completion was empty.

The base model scored **76.2% (125/164)** with the same eval script. The SFT model itself learned correctly (loss 0.94, token accuracy 78%) — the 0% is purely a config bug.

## Root Cause

TRL's SFTTrainer sets `tokenizer.pad_token = tokenizer.eos_token` when no pad token exists. For Nemotron, `eos_token` is `<|im_end|>` (token 12), which is also in the `eos_token_id` list `[2, 11, 12]`. The trainer saves this collision into `generation_config.json`.

`GenerationConfig.validate()` checks various parameter combinations (do_sample vs top_p, etc.) but does **not** check for pad/eos collision.

## Proposed Fix

Add a validation check in `GenerationConfig.validate()`:

```python
if self.pad_token_id is not None and self.eos_token_id is not None:
    eos_ids = self.eos_token_id if isinstance(self.eos_token_id, list) else [self.eos_token_id]
    if self.pad_token_id in eos_ids:
        warnings.warn(
            f"pad_token_id ({self.pad_token_id}) is in eos_token_id list ({eos_ids}). "
            f"This may cause model.generate() to stop immediately. "
            f"Consider setting pad_token_id to a token not in eos_token_id."
        )
```

This would warn users at config load time, before they waste hours running evals that produce 0%.

## Environment

- transformers: 5.16.0.dev0 (main branch)
- trl: 1.10.0
- Model: nvidia/NVIDIA-Nemotron-Nano-9B-v2 (NemotronH hybrid Mamba-2 + Attention)
- GPU: NVIDIA GB10 (119GB VRAM)

## Related

- TRL issue #1384: "SFTTrainer should not set tokenizer.pad_token_id = tokenizer.eos_token_id" (open since 2023)
- TRL PR #3200: Fix SFT masking EOS when equal to PAD (merged — fixes training, but not the saved config)
- Full writeup: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

### PR DESCRIPTION
**Pull request by davidnichols-ops**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48022) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48022)

## What does this PR do?

Adds a validation check in `GenerationConfig.validate()` that warns when `pad_token_id` is also present in the `eos_token_id` list. This addresses [#48016](https://github.com/huggingface/transformers/issues/48016).

## The problem

When `pad_token_id` overlaps with one of the tokens in `eos_token_id`, `model.generate()` can stop immediately because every padded position is treated as a finished sequence. This produces empty or degenerate outputs with no obvious error message, making it very hard to diagnose.

In our case, an SFT model was configured with `pad_token_id=12` and `eos_token_id=[2, 11, 12]`. Because token `12` was both the pad token and an EOS token, generation terminated instantly, and the model scored **0% on HumanEval**. There was no warning or error to indicate why.

A full writeup of the debugging process is available here: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

## The fix

This PR adds a check in the `validate()` method (section 2.7, "other incorrect combinations") that detects when `pad_token_id` is in the `eos_token_id` list. It follows the existing `minor_issues` reporting pattern:

- **Default mode**: logs a warning via `logger.warning_once` with a short summary, and the full details at INFO level — consistent with all other validation checks in the method.
- **Strict mode** (`strict=True`): raises a `ValueError`, consistent with how all minor issues are escalated.

The check handles both `eos_token_id` as a list and as a single int, and is a no-op when either `pad_token_id` or `eos_token_id` is `None`.

## Before

```python
from transformers import GenerationConfig
config = GenerationConfig(pad_token_id=12, eos_token_id=[2, 11, 12])
config.validate()
# No warning — silently misconfigured
```

## After

```python
from transformers import GenerationConfig
config = GenerationConfig(pad_token_id=12, eos_token_id=[2, 11, 12])
config.validate()
# WARNING: The following generation flags are not valid and may be ignored: ["pad_token_id"].
# INFO: `pad_token_id` (12) is in `eos_token_id` ([2, 11, 12]). This may cause `generate` to
#       stop immediately, as every padded position is treated as a finished sequence.
#       Consider setting `pad_token_id` to a token that is not in `eos_token_id`.
```

## Tests

Added `test_validate_pad_token_id_in_eos_token_id` in `tests/generation/test_configuration_utils.py` covering:
- `pad_token_id` in `eos_token_id` list → warning
- `pad_token_id` equals single int `eos_token_id` → warning
- `pad_token_id` not in `eos_token_id` → no warning
- `pad_token_id` is `None` → no warning
- `eos_token_id` is `None` → no warning
- `strict=True` → `ValueError`

Closes #48016

### CODE REVIEW

**Comment by davidnichols-ops** [comment]
CI failures on the latest run are all GitHub infrastructure issues (429 Too Many Requests, 502 Bad Gateway, 503 Service Unavailable when downloading GitHub Actions). No actual test failures from our changes. Re-triggering CI.

**Comment by davidnichols-ops** [comment]
Not trying to be loud here, but this has been sitting with no feedback for almost 3 weeks. If someone could review this, (even a peer, contributor) would be greatly appreciated.

**Comment by davidnichols-ops** [comment]
Closing this after re-examining the behavior and the appropriate layer for the check.

The current approach has three problems:

1. It warns on valid, mainstream configurations. For example, `Qwen/Qwen2.5-0.5B-Instruct` ships with `pad_token_id=151643` included in `eos_token_id=[151645, 151643]`, so merely loading that configuration would produce the warning.

2. The message describes the wrong mechanism. `EosTokenCriteria` examines newly generated tokens, not padded prompt positions. In a runtime reproduction with `pad_token_id` included in `eos_token_id`, generation produced the full requested token budget rather than stopping immediately. The actual concern is that this configuration prevents automatic attention-mask inference.

3. `GenerationConfig.validate()` cannot see the facts that determine whether the configuration is harmful: whether the input is batched and whether an attention mask was supplied. Transformers already has a runtime warning at the point where those facts are available.

Current TRL source also shows how this configuration can arise: `SFTTrainer` falls back to the tokenizer’s EOS token when no padding token exists and copies the resulting padding ID into the model’s generation configuration. Transformers’ trainer alignment performs a similar copy. That producer and persistence behavior needs separate investigation before deciding whether either project should change it.

A narrower Transformers improvement could clarify the existing runtime warning so it explicitly names the `pad_token_id ∈ eos_token_id` collision. I would open that separately if maintainers want it.

Thanks for the direction on #48016. I’m closing this implementation because a configuration-load warning is the wrong layer for the problem.
